### PR TITLE
(WIP) (PUP-6496) Add support for using HOCON for main config file

### PR DESCRIPTION
**Pull request by janelu2**
This PR adds  a backward-compatible way to allow for users to use a `hiera.conf` rather than `hiera.yaml` file. If the `hiera.conf` file is found, then it will parse it as hocon, otherwise it will use the `hiera.yaml` file. This PR also adds some tests to make sure that if given two configuration files, that it would read `hiera.conf` rather than `hiera.yaml`.

### CODE REVIEW

**Comment by janelu2** [comment]
I will rebase my branch after it passes the appV tests.. I don't want to re-push and restart the appV tests while it's already been running for almost an hour now.

**Comment by janelu2** [comment]
@cprice404 @thallgren let me know if there are other tests or anything else that I should do/add to this PR.

**Comment by cprice404** [comment]
@janelu2 so is 4ea2897 the only commit for review here?

**Comment by janelu2** [comment]
@cprice404 yisss

**Comment by janelu2** [comment]
@cprice404 added the other test, it's the stuff under hiera_hocon_no_yaml_config, also the test has been updated in the hiera_data_provider_spec.rb

File: spec/unit/data_providers/hiera_data_provider_spec.rb [hunk 1]
```diff
@@ -58,11 +58,16 @@ def extract_notifications(catalog)
     expect(resources).to include('module data param_a is 100, module data param_b is 200, module data param_c is 300, module data param_d is 400, module data param_e is 500, module data param_f is 600, module data param_g is 700')
   end
 
-  it 'uses ' do
+  it 'reads hirea.conf and parses it as hocon' do
     resources = compile_and_get_notifications('hiera_defaults')
     expect(resources).to include('module data param_a is 100, param default is 200, env data param_c is 300')
   end
 
+  it 'loads and parses heira.conf even if there is no hiera.yaml' do
```
Comment (janelu2) [inline spec/unit/data_providers/hiera_data_provider_spec.rb:66]: @cprice404 this is the new test

File: spec/unit/data_providers/hiera_data_provider_spec.rb [hunk 2]
```diff
@@ -58,11 +58,16 @@ def extract_notifications(catalog)
     expect(resources).to include('module data param_a is 100, module data param_b is 200, module data param_c is 300, module data param_d is 400, module data param_e is 500, module data param_f is 600, module data param_g is 700')
   end
 
-  it 'uses ' do
+  it 'reads hirea.conf and parses it as hocon' do
```
Comment (cprice404) [inline spec/unit/data_providers/hiera_data_provider_spec.rb]: typo, `hirea`

File: spec/unit/util/windows/sid_spec.rb [hunk 3]
```diff
@@ -99,6 +99,7 @@
         # Ruby seems to use the local codepage when making COM calls
         # if this fails, might want to use Windows API directly instead to ensure bytes
         user = Puppet::Util::Windows::ADSI.create(username, 'user')
+        user.SetPassword('PUPPET_RULeZ_123!')
```
Comment (cprice404) [inline spec/unit/util/windows/sid_spec.rb:102]: What's up with this line?

Comment (janelu2) [inline spec/unit/util/windows/sid_spec.rb:102]: it's from a different commit, I'm going to rebase my branch that this PR is based off of so it goes away

Comment (cprice404) [inline spec/unit/util/windows/sid_spec.rb:102]: gotcha, thanks.

File: lib/puppet/data_providers/hiera_config.rb [hunk 4]
```diff
@@ -58,12 +59,20 @@ def self.create_config_type
     # @api public
     def initialize(config_root)
       @config_root = config_root
-      @config_path = config_root + 'hiera.yaml'
-      if @config_path.exist?
-        @config = validate_config(HieraConfig.symkeys_to_string(YAML.load_file(@config_path)))
+      config_path_hocon = config_root + 'hiera.conf'
+      config_path_yaml = config_root + 'hiera.yaml'
+      if config_path_hocon.exist?
+        @config = validate_config(HieraConfig.symkeys_to_string(Hocon.load(File.absolute_path(config_path_hocon))))
```
Comment (thallgren) [inline lib/puppet/data_providers/hiera_config.rb]: @Does Hocon#load produce a hash with symbols? Unless it does, the `#symkeys_to_string` call is unnecessary.

Comment (janelu2) [inline lib/puppet/data_providers/hiera_config.rb]: From [these docs](https://github.com/typesafehub/config), `load`returns a `Config` object which is an immutable map.

Comment (janelu2) [inline lib/puppet/data_providers/hiera_config.rb]: I'm going to remove that call, it is unnecessary.

**Comment by cprice404** [comment]
Hi Jane hope you are well! :)

**Comment by janelu2** [comment]
@cprice404 Hi Chris!! Sorry I meant to respond sooner. Yes I’m doing well :) I’ve finally left amazon. What a rollercoaster that was. How are you doing? I see you’ve left also. We need to catch up soon, the past years have been too busy for me to do anything but overwork hahaha.

**Comment by cprice404** [comment]
if you are still in portland we should get some folks together for lunch sometime.  i'll ping you on linkedin or something :)

**Pull request closed without merging**